Ast/src/Parser.cpp

```diff
@@ -897,6 +916,26 @@ AstStat* Parser::parseTypeAlias(const Location& start, bool exported)
     return allocator.alloc<AstStatTypeAlias>(Location(start, type->location), name->name, name->location, generics, genericPacks, type, exported);
 }
 
+// type function Name `(' arglist `)' `=' funcbody `end'
+AstStat* Parser::parseTypeFunction(const Location& start)
+{
+    Lexeme matchFn = lexer.current();
+    nextLexeme();
+
+    // parse the name of the type function
+    std::optional<Name> fnName = parseNameOpt("type function name");
+    if (!fnName)
+        fnName = Name(nameError, lexer.current().location);
+
+    matchRecoveryStopOnToken[Lexeme::ReservedEnd]++;
+
+    AstExprFunction* body = parseFunctionBody(/* hasself */ false, matchFn, fnName->name, nullptr, AstArray<AstAttr*>({nullptr, 0})).first;
```

Comment:
Also since the parser does name resolutions for locals, this means type functions can have upvalues which is strange.

```
local x = 5
type function f()
  print(x) --> x = 5 here, should be unknown global
end
```